phoenix-load-balancer/src/main/java/org/apache/phoenix/loadbalancer/service/LoadBalancer.java

```diff
@@ -0,0 +1,165 @@
+            if ((curaFramework != null) && (unhandledErrorListener != null)){
+                curaFramework.getUnhandledErrorListenable()
+                        .removeListener(unhandledErrorListener);
+            }
+            for (Closeable closeable : closeAbles) {
+                CloseableUtils.closeQuietly(closeable);
+            }
+        }
+    }
+
+    /**
+     * Return Singleton Load Balancer every single time.
+     * @return LoadBalancer
+     */
+    public static LoadBalancer getLoadBalancer() {
+        return loadBalancer;
+    }
+
+    /**
+     * return the location of Phoenix Query Server
+     * in form of String. The method return a random PQS server
+     * from the cache.
+     * @return - PQS Location ( host:port)
+     * @throws Exception
+     */
+    public PhoenixQueryServerNode getSingleServiceLocation()  throws Exception{
+        List<PhoenixQueryServerNode> childNodes = conductSanityCheckAndReturn();
+        // get an random connect string
+        int i = ThreadLocalRandom.current().nextInt(0, childNodes.size());
+        return childNodes.get(i);
+    }
+
+    /**
+     * return all the location of Phoenix Query Server from cache
+     * @return - PQS Locations ( host1:port1,host2:port2)
+     * @throws Exception
+     */
+    public List<PhoenixQueryServerNode> getAllServiceLocation()  throws Exception{
+        return conductSanityCheckAndReturn();
+    }
+
+    private List<PhoenixQueryServerNode> conductSanityCheckAndReturn() throws Exception{
+        Preconditions.checkNotNull(curaFramework
+                ," curator framework in not initialized ");
+        Preconditions.checkNotNull(cache," cache value is not initialized");
+        boolean connected = curaFramework.getZookeeperClient().isConnected();
+        if (!connected) {
+            String message = " Zookeeper seems to be down. The data is stale ";
+            Exception exception =
+                    new Exception(message);
+            LOG.error(message, exception);
+            throw exception;
+        }
+        List<String> currentNodes = curaFramework.getChildren().forPath(CONFIG.getParentPath());
+        List<PhoenixQueryServerNode> returnNodes = new ArrayList<>();
+        for(String node:currentNodes) {
+            byte[] bytes = curaFramework.getData().forPath(CONFIG.getParentPath() + "/" + node);
+            returnNodes.add(PhoenixQueryServerNode.fromJsonString(new String(
+                    bytes, StandardCharsets.UTF_8)));
+        }
+        return returnNodes;
+    }
+    private String getZkConnectString(){
+        return CONFIG.getZkConnectString();
+    }
+
+    private ConnectionStateListener getConnectionStateListener(){
+        return new ConnectionStateListener() {
+            @Override
+            public void stateChanged(CuratorFramework client, ConnectionState newState) {
+                if (!newState.isConnected()) {
+                    LOG.error(" connection to zookeeper broken ");
```

Comment:
Please include the connection state. ZK has a complicated graph of connectivity states (many of which are recoverable). To debug this, we would want to know what state we moved into.